### DISCUSSION TITLE
(Newton's Law of) Universal Gravitation in MuJoCo

### DISCUSSION THREAD

**Discussion by handsomir** [google-deepmind/mujoco#976] [General]
Aloha MuJoCo Team,

I am learning MuJoCo with the goal of developing and testing space robotics sims. Right now we are using a force-free environment to set up our sims and test configurations. The next step is to introduce external force models starting with gravity. From the MuJoCo XML and API documentation I gather that a straightforward way to implement Newton's law of universal gravitation is to:

1. Change gravity: real(3), from “0 0 -9.81” to "0 0 0" and
2. Write an external module that returns the solution of $\ddot{r} = - \frac{\mu}{|r|^3}  r + a_T$ together with six initial conditions, where $r$, real(3), is the Cartesian position vector, $\ddot{r}$, real(3), is its second order derivative w/respect to time, $\mu$ is the gravitational parameter of the central body, and $a_T$, real(3), is the acceleration due to the thrust in reference frame in which $r$ is expressed.
3. Calculate the force on each of the $n$ links, $F_i=m_i \ddot{r}_i, i \in \\{ 1, ... ,n \\}$ and add it to the respective mjcb_passive

I have done a scan of the discussions and haven't found anything similar. However, if someone has worked on something similar and posted about it please let me know. I appreciate feedback and suggestions.

Regards,

BU

**Comment by yuvaltassa** [comment]
Yup, multi-body Newtonian gravitation should be easy to implement as a passive force callback or plugin. Looking forward to seeing some cool animations 🙂

**Reply by handsomir** [reply]
Thank you @yuvaltassa . I'm sure we'll have additional questions. We'll post results when we are confident in our gravity model. BTW, I saw one of your replies, in a thread on actuators, that "_nature is always second order_."  I also used to tell the students in my controls class that "_mother nature always uses a double integrator_." =D

**Reply by yuvaltassa** [reply]
On a personal note, if you are not familiar with [Osmos](https://apps.apple.com/gb/app/osmos/id382991304), I highly recommend it. It's an interactive game using Newtonian gravity.

**Reply by saran-t** [reply]
Only _some_ levels in Osmos actually have gravitating motes :)

**Reply by handsomir** [reply]
Thanks - I will take a look at Osmos. I wonder if it's inspired by "The Mote in God's Eye" by Niven and Pournelle.

**Comment by tmplt** [comment]
I am also interested in this. Any updates, @TenguBoss?

**Reply by handsomir** [reply]
@tmplt we have decided to work in "zero gee" until I become proficient with MuJoCo and its API. I plan to get back to the Newtonian gravity model this spring. I also plan to work on including variable mass in space vehicle sims, similar to the [discussion here](https://github.com/google-deepmind/mujoco/discussions/914)

**Comment by tmplt** [comment]
I'm trying my hand at this via callback, but I'm not sure how to install it. In a REPL:
```python
>>> import mujoco
>>> def callback(model, data):
...     pass
... 
>>> mujoco.set_mjcb_passive(callback)
>>> mujoco.MjModel.from_xml_string('<mujoco/>')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Error: engine error: Python exception raised
```

Could you share an example callback install, @yuvaltassa? I can't find one in documentation. Thanks.

**Reply by yuvaltassa** [reply]
The documentation is [here](https://mujoco.readthedocs.io/en/latest/python.html#callbacks), there are tests [here](https://github.com/google-deepmind/mujoco/blob/main/python/mujoco/bindings_test.py)

**Reply by tmplt** [reply]
I believe [I hit an edge case](https://github.com/google-deepmind/mujoco/issues/1376). Setting the passive force *after* the model has been created works as per documentation.
